File: 백준/Silver/2805. 나무 자르기/나무 자르기.py
from sys import stdin
n,m = map(int, stdin.readline().split())
data = list(map(int, stdin.readline().split()))
start=0
end=max(data)
result=0
while start<=end:
	total=0
	mid=(start+end)//2
	for i in data:
		if i > mid:
			total+=i-mid
	if total < m:
		end = mid-1
	else:
		result = mid
		start = mid+1
print(result)

File: 백준/Silver/2805. 나무 자르기/나무 자르기.js
const fs = require('fs');
const filePath = process.platform === 'linux' ? '/dev/stdin' : './input.txt';
const input = require('fs').readFileSync(filePath).toString().split('\n')

const target = input[0].split(' ').map(Number)[1];
const arr = input[1].split(' ').map(Number);

let start = 0;
let end = Math.max(...arr);

let result = 0
while(start<=end){
  let mid = parseInt((start + end)/2);
  let total = 0
  for( x of arr){
    if(x > mid){
      total += x-mid
    }
  }
  if(total < target){
    end = mid -1
  }else {
    result = mid;
    start = mid +1
  }
}
console.log(result);
